src/appleseed-max-impl/utilities.cpp

```diff
@@ -57,33 +57,233 @@
 namespace asf = foundation;
 namespace asr = renderer;
 
+std::string wide_to_utf8(const std::wstring& wstr)
+{
+    if (wstr.empty())
+        return std::string();
+
+    const int wstr_size = static_cast<int>(wstr.size());
+    const int result_size = WideCharToMultiByte(CP_UTF8, 0, &wstr[0], wstr_size, nullptr, 0, nullptr, nullptr);
+
+    std::string result(result_size, 0);
+    WideCharToMultiByte(CP_UTF8, 0, &wstr[0], wstr_size, &result[0], result_size, nullptr, nullptr);
+
+    return result;
+}
+
+std::string wide_to_utf8(const wchar_t* wstr)
+{
+    const int result_size = WideCharToMultiByte(CP_UTF8, 0, wstr, -1, nullptr, 0, nullptr, nullptr);
+
+    std::string result(result_size - 1, 0);
+    WideCharToMultiByte(CP_UTF8, 0, wstr, -1, &result[0], result_size - 1, nullptr, nullptr);
+
+    return result;
+}
+
+std::wstring utf8_to_wide(const std::string& str)
+{
+    if (str.empty())
+        return std::wstring();
+
+    const int str_size = static_cast<int>(str.size());
+    const int result_size = MultiByteToWideChar(CP_UTF8, 0, &str[0], str_size, nullptr, 0);
+
+    std::wstring result(result_size, 0);
+    MultiByteToWideChar(CP_UTF8, 0, &str[0], str_size, &result[0], result_size);
+
+    return result;
+}
+
+std::wstring utf8_to_wide(const char* str)
+{
+    const int result_size = MultiByteToWideChar(CP_UTF8, 0, str, -1, nullptr, 0);
+
+    std::wstring result(result_size - 1, 0);
+    MultiByteToWideChar(CP_UTF8, 0, str, -1, &result[0], result_size - 1);
+
+    return result;
+}
+
+std::string get_root_path()
+{
+    wchar_t path[MAX_PATH];
+    const DWORD path_length = sizeof(path) / sizeof(wchar_t);
+
+    const auto result = GetModuleFileName(g_module, path, path_length);
+    DbgAssert(result != 0);
+
+    PathRemoveFileSpec(path);
+
+    return wide_to_utf8(path);
+}
+
+WStr replace_extension(const WStr& file_path, const WStr& new_ext)
+{
+    const int i = file_path.last(L'.');
+    WStr new_file_path = i == -1 ? file_path : file_path.Substr(0, i);
+    new_file_path.Append(new_ext);
+    return new_file_path;
+}
+
+bool is_bitmap_texture(Texmap* map)
+{
+    if (map == nullptr)
+        return false;
+
+    if (map->ClassID() != Class_ID(BMTEX_CLASS_ID, 0))
+        return false;
+
+    Bitmap* bitmap = static_cast<BitmapTex*>(map)->GetBitmap(0);
+
+    if (bitmap == nullptr)
+        return false;
+
+    return true;
+}
+
+bool is_supported_procedural_texture(Texmap* map)
+{
+    if (map == nullptr)
+        return false;
+
+    auto part_a = map->ClassID().PartA();
+    auto part_b = map->ClassID().PartB();
+
+    switch (part_a)
+    {
+      case 0x64035FB9:              // Tiles.
```

Comment:
Those comments should be without the periods probably?

Reply:
Good point!